kernel/ps2/src/lib.rs

```diff
@@ -1,484 +1,704 @@
+    input_buffer_full: bool1,
+    /// Cleared on reset; set when the system passed Power-on self-test
+    #[allow(dead_code)]
+    system_passed_self_test: bool1,
+    /// Input buffer should be written to: 0 - [PS2_DATA_PORT], 1 - [PS2_COMMAND_AND_STATUS_PORT]
+    #[allow(dead_code)]
+    input_buffer_is_command: u1,
+    /// Whether or not communication is inhibited (via switch) / Unknown (chipset specific) / (more likely unused on modern systems)
+    #[allow(dead_code)]
+    keyboard_enabled: bool1,
+    ///// Keyboard didn't generate clock signals within 15 ms of "request-to-send" (exclusive to AT-compatible mode)
+    // transmit_timeout: bool1,
+    ///// Keyboard didn't generate clock signals within 20 ms of command reception (exclusive to AT-compatible mode)
+    // receive_timeout: bool1,
+    /// Similar to `output_buffer_full`, except for mouse
+    #[allow(dead_code)]
+    mouse_output_buffer_full: bool1,
+    /// Timeout during keyboard command receive or response (Same as [transmit_timeout] + [receive_timeout])
+    #[allow(dead_code)]
+    timeout_error: bool1,
+    /// Should be odd parity, set to `true` if even parity received
+    #[allow(dead_code)]
+    parity_error: bool1,
+}
+
+/// Read the PS/2 status port/register
+fn status_register() -> ControllerToHostStatus {
+    ControllerToHostStatus::from_bytes([PS2_COMMAND_AND_STATUS_PORT.lock().read()])
+}
+
+//NOTE: if we are _not_ in an interrupt handler, it might be necessary to check for `output_buffer_full` first
+/// Read data from the PS/2 data port.
+/// This is the lowest level interface to the port.
+fn read_data() -> u8 {
+    PS2_DATA_PORT.lock().read()
+}
+
+/// All types of data writable to the [PS2_DATA_PORT].
+enum WritableData {
+    Configuration(ControllerConfigurationByte),
+    HostToDevice(HostToDevice)
+}
+
+// NOTE: could also use a macro / reminds me of serde
+impl From<WritableData> for u8 {
+    fn from(value: WritableData) -> Self {
+        match value {
+            WritableData::Configuration(value) => value.into_bytes()[0],
+            WritableData::HostToDevice(value) => {
+                match value {
+                    HostToDevice::Keyboard(value) => {
+                        match value {
+                            HostToKeyboardCommandOrData::KeyboardCommand(c) => c as u8,
+                            HostToKeyboardCommandOrData::LEDState(l) => l.into_bytes()[0],
+                            HostToKeyboardCommandOrData::ScancodeSet(s) => s as u8,
+                        }
+                    }
+                    HostToDevice::Mouse(value) => {
+                        match value {
+                            HostToMouseCommandOrData::MouseCommand(c) => c as u8,
+                            HostToMouseCommandOrData::MouseResolution(r) => r as u8,
+                            HostToMouseCommandOrData::SampleRate(s) => s as u8,
+                        }
+                    }
+                }
+            }
```

Comment:
Can reduce indentation and improve legibility like so:

```suggestion
            WritableData::Configuration(value) => value.into_bytes()[0],
            WritableData::HostToDevice(value) => match value {
                HostToDevice::Keyboard(value) => match value {
                    HostToKeyboardCommandOrData::KeyboardCommand(c) => c as u8,
                    HostToKeyboardCommandOrData::LEDState(l) => l.into_bytes()[0],
                    HostToKeyboardCommandOrData::ScancodeSet(s) => s as u8,
                }
                HostToDevice::Mouse(value) => match value {
                        HostToMouseCommandOrData::MouseCommand(c) => c as u8,
                        HostToMouseCommandOrData::MouseResolution(r) => r as u8,
                        HostToMouseCommandOrData::SampleRate(s) => s as u8,
                    }
                }
            }
```